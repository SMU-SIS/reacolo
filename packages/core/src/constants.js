/**
 * Symbol used for wild case in the matcher syntax.
 * @private
 */
export const WILD_CASE = '*';

/**
 * Context key of the model provided by the provider.
 * @private
 */
export const MODEL_CONTEXT_KEY = 'reacoloModel';

/**
 * Event emitted by the model when it has been updated.
 * @private
 */
export const MODEL_UPDATE_EVENT = 'reacolo:model:update';
