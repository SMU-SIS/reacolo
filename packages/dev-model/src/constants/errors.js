/**
 * @module reacolo-dev-model/constants/errors
 * @private
 */

/**
 * @type {Error}
 * @alias module:reacolo-dev-model.NotConnectedError
 */
export class NotConnectedError extends Error {}

/**
 * @type {Error}
 * @alias module:reacolo-dev-model.RequestTimeoutError
 */
export class RequestTimeoutError extends Error {}

/**
 * @type {Error}
 * @alias module:reacolo-dev-model.RequestFailedError
 */
export class RequestFailedError extends Error {}

/**
 * @type {Error}
 * @alias module:reacolo-dev-model.AlreadyConnectedError
 */
export class AlreadyConnectedError extends Error {}
