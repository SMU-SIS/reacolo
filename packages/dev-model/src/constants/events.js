/**
 * @module reacolo-dev-model/constants/events
 * @private
 */

/**
 * Event sent on model's updates.
 * @type {'reacolo:model:update'}
 * @alias module:reacolo-dev-model.MODEL_UPDATE_EVT
 */
export const MODEL_UPDATE_EVT = 'reacolo:model:update'; // eslint-disable-line import/prefer-default-export
