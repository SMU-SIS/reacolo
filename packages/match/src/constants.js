/**
 * Symbol used for wild case in the matcher syntax.
 * @private
 */
export const WILD_CASE = '*'; // eslint-disable-line import/prefer-default-export
