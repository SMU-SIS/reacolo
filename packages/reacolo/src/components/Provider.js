import { Component } from 'react';
import propTypes from 'prop-types';
import { modelPropType } from '../utils/prop-types';

export default class Provider extends Component {
  constructor(props) {
    super(props);
    this.model = props.model;
  }
  getChildContext() {
    return { model: this.model };
  }
  componentWillReceiveProps(props) {
    if (props.model !== this.model) {
      throw new Error('<Provider> does not support changing the model.');
    }
  }
  render() {
    return this.props.children;
  }
}

Provider.childContextTypes = {
  // eslint-disable-next-line react/no-typos
  model: modelPropType.isRequired,
};

Provider.propTypes = {
  // eslint-disable-next-line react/no-typos
  model: modelPropType.isRequired,
  children: propTypes.element.isRequired,
};
