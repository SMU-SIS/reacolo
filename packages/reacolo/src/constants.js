/**
 * Symbol used for wild case in the matcher syntax.
 */
export const WILD_CASE = '.';

/**
 * Context key of the model provided by the provider.
 */
export const MODEL_CONTEXT_KEY = 'reacoloModel';
