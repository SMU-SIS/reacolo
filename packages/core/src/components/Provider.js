import { Component } from 'react';
import propTypes from 'prop-types';
import { modelPropType } from '../utils';
import { MODEL_CONTEXT_KEY } from '../constants';

/**
 * Provides the reacolo model as a context property.
 */
export default class Provider extends Component {
  constructor(props) {
    super(props);
    this.model = props.model;
  }
  getChildContext() {
    return { [MODEL_CONTEXT_KEY]: this.model };
  }
  componentWillReceiveProps(props) {
    if (props.model !== this.model) {
      throw new Error('<Provider> does not support changing the model.');
    }
  }
  render() {
    return this.props.children;
  }
}

Provider.childContextTypes = {
  // eslint-disable-next-line react/no-typos
  [MODEL_CONTEXT_KEY]: modelPropType.isRequired,
};

Provider.propTypes = {
  // eslint-disable-next-line react/no-typos
  model: modelPropType.isRequired,
  children: propTypes.element.isRequired,
};
