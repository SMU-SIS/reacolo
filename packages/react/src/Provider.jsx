import React, { Component, Fragment } from 'react';
import propTypes from 'prop-types';
import { modelPropType } from './utils';
import { MODEL_CONTEXT_KEY } from './constants';

/**
 * Provides the reacolo model as a context property. To be used with <Context>,
 * <Switch> and connect.
 */
export default class Provider extends Component {
  constructor(props) {
    super(props);
    this.model = props.model;
  }
  getChildContext() {
    return { [MODEL_CONTEXT_KEY]: this.model };
  }
  componentWillReceiveProps(props) {
    if (props.model !== this.model) {
      throw new Error('<Provider> does not support changing the model.');
    }
  }
  render() {
    return <Fragment>{this.props.children}</Fragment>;
  }
}

Provider.childContextTypes = {
  // eslint-disable-next-line react/no-typos
  [MODEL_CONTEXT_KEY]: modelPropType.isRequired,
};

Provider.propTypes = {
  // eslint-disable-next-line react/no-typos
  model: modelPropType.isRequired,
  children: propTypes.element.isRequired,
};
