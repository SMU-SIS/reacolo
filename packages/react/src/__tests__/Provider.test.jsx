import React, { Component } from 'react';
import propTypes from 'prop-types';
import Enzyme, { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import Provider from '../Provider.jsx';
import { MODEL_CONTEXT_KEY } from '../constants';

Enzyme.configure({ adapter: new EnzymeAdapter() });

describe('Provider', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error');
  });

  // Mock the console.
  const mockConsole = () => {
    global.console.error.mockImplementation(() => {});
  };

  // Make sure the console is restored after each tests.
  afterEach(() => {
    global.console.error.mockRestore();
  });

  // Child cannot be a stateless function or findRenderedComponentWithType
  // does not work.
  // eslint-disable-next-line react/prefer-stateless-function
  class Child extends Component {
    render() {
      return (
        <div>
          reacolo model name: {this.context[MODEL_CONTEXT_KEY].mockName}
        </div>
      );
    }
  }
  Child.contextTypes = {
    [MODEL_CONTEXT_KEY]: propTypes.object.isRequired,
  };

  const createModel = props => ({
    addListener: jest.fn(),
    removeListener: jest.fn(),
    getContext: jest.fn(),
    getStore: jest.fn(),
    mockName: 'mock-model',
    ...props,
  });

  it("should add the store to the child's context", () => {
    const model = createModel();
    const tree = mount(
      <Provider model={model}>
        <Child />
      </Provider>,
    );
    expect(toJSON(tree)).toMatchSnapshot();
  });

  it('should throw if the store is changed', () => {
    // Avoid React's error message being thrown on jest's console.
    mockConsole();

    const model1 = createModel({ mockName: 'mock-model-1' });
    const model2 = createModel({ mockName: 'mock-model-2' });

    const tree = mount(
      <Provider model={model1}>
        <Child />
      </Provider>,
    );

    expect(() => {
      tree.setProps({ model: model2 });
    }).toThrow();

    expect(toJSON(tree)).toMatchSnapshot();
  });
});
