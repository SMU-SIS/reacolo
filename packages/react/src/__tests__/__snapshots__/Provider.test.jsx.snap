// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider should add the store to the child's context 1`] = `
<Provider
  model={
    Object {
      "addListener": [MockFunction],
      "getContext": [MockFunction],
      "getStore": [MockFunction],
      "mockName": "mock-model",
      "removeListener": [MockFunction],
    }
  }
>
  <Child>
    <div>
      reacolo model name: 
      mock-model
    </div>
  </Child>
</Provider>
`;

exports[`Provider should throw if the store is changed 1`] = `
<Provider
  model={
    Object {
      "addListener": [MockFunction],
      "getContext": [MockFunction],
      "getStore": [MockFunction],
      "mockName": "mock-model-1",
      "removeListener": [MockFunction],
    }
  }
>
  <Child>
    <div>
      reacolo model name: 
      mock-model-1
    </div>
  </Child>
</Provider>
`;
